analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ObjectsShouldNotBeDisposedMoreThanOnceBase.cs

```diff
@@ -18,10 +18,41 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
 
 public abstract class ObjectsShouldNotBeDisposedMoreThanOnceBase : SymbolicRuleCheck
 {
     protected const string DiagnosticId = "S3966";
-    protected const string MessageFormat = "Refactor this code to make sure '{0}' is disposed only once.";
+    protected const string MessageFormat = "Resource '{1}' {0}";
+    protected const string MessageDisposeTwice = "has already been disposed. Refactor the code to dispose it only once.";
+    protected const string MessageDisposeUsing = "is ensured to be disposed by this using statement. You don't need to dispose it twice.";
+
+
+    private static readonly string[] DisposeMethods = { "Dispose", "DisposeAsync", "Close" };
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        if (context.Operation.Instance.AsInvocation() is { } invocation && DisposeMethods.Contains(invocation.TargetMethod.Name))
+        {
+            if (state[invocation.Instance]?.HasConstraint(DisposableConstraint.Disposed) is true)
+            {
+                ReportIssue(context.Operation.Instance, IsPartOfUsingStatement(invocation) ? MessageDisposeUsing : MessageDisposeTwice, invocation.Instance.Syntax.ToString());
```

Comment:
Can I report the issue on the symbol?
Currently it's on the invocation.

Reply:
Wrong words- Can I report on the Instance of the `invocation.Instance` - whatever this might be (expression, object instance etc).

Reply:
I still don't get it, the reporting is on `Operation.Instance` already. Try to write a test case with `^^^^` here to explain the question please

Reply:
Two examples

Now it's:
```csharp
b.Dispose(); // Noncompliant {{Resource 'b' has already been disposed explicitly or implicitly through a using statement. Please remove the redundant disposal.}}
^^^^^^^^^^^
```

Can it be?
```csharp
b.Dispose(); // Noncompliant {{Resource 'b' has already been disposed explicitly or implicitly through a using statement. Please remove the redundant disposal.}}
^
```

Reply:
PS: I think where we raise now makes sense - that's for me to understand.

Reply:
The first ReportIssue argument decides the location. Use
```
ReportIssue(invocation.Instance, invocation.Instance.Syntax.ToString());
```
instead of 
```
ReportIssue(context.Operation.Instance, invocation.Instance.Syntax.ToString());
```

Reply:
Thanks!